### ISSUE TITLE
Contest Processing

### ISSUE THREAD

**Issue by ccev** [UnownHash/Golbat#125]
it might be cool to have proper contest support. here's the needed data:

GMO -> forts -> PokestopIncidentDisplayProto
```
pokestop_displays {
    incident_id "-1601456624153831644"
    incident_start_ms 1688626800000
    incident_expiration_ms 1688666400001
    incident_display_type IncidentDisplayType.INCIDENT_DISPLAY_TYPE_INCIDENT_CONTEST:9
}
```

GetContestDataOutProto
```
METHOD_GET_CONTEST_DATA | 2105

GetContestDataProto {
    fort_id "ed8a635917594ef2bc25b36993b2dbf0.16"
}

GetContestDataOutProto {
    status Status.SUCCESS:1
    
    contest_incident {
        
        contests {
            contest_id "ed8a635917594ef2bc25b36993b2dbf0.16-CUSTOM-MAX-POKEMON_SIZE-1688630400000"
            
            focus {
                
                pokemon {
                    pokedex_id HoloPokemonId.SQUIRTLE:7
                }
            }
            
            metric {
                pokemon_metric ContestPokemonMetric.POKEMON_SIZE:1
                ranking_standard ContestRankingStandard.MAX:2
            }
            
            schedule {
                
                contest_cycle {
                    start_time_ms 1688630400000
                    end_time_ms 1688666400000
                    contest_occurrence ContestOccurrence.CUSTOM:8
                    custom_cycle_warmup_duration_ms 3600000
                    custom_cycle_cooldown_duration_ms 1
                }
            }
        }
    }
}

```

GetPokemonSizeContestEntryProto
```
METHOD_GET_POKEMON_SIZE_CONTEST_ENTRY | 2104

GetPokemonSizeContestEntryProto {
    contest_id "ed8a635917594ef2bc25b36993b2dbf0.16-CUSTOM-MAX-POKEMON_SIZE-1688630400000"
    entry_count 2147483647
    
    contest_metric {
        pokemon_metric ContestPokemonMetric.POKEMON_SIZE:1
        ranking_standard ContestRankingStandard.MAX:2
    }
}

GetPokemonSizeContestEntryOutProto {
    status Status.SUCCESS:1
    total_entries 13
    
    contest_entries {
        pokedex_id HoloPokemonId.SQUIRTLE:7
        score 900.9230982058893
        rank 1
        trainer_name "Inzeckt"
        team Team.TEAM_BLUE:1
        pokemon_id 16024912515536508900
        
        pokemon_display {
            costume Costume.JAN_2020_NOEVOLVE:11
            gender Gender.FEMALE:2
            form Form.SQUIRTLE_NORMAL:181
            display_id 2544955071119492613
        }
        
        player_avatar {
            ...
        }
    }
    
    ... (contest_entries is a repeated field)
}

```

we'll probably need a new table `contest_entries` that contains the participants. if a contest finishes, the entries should get removed. contests themselves can just be treated as as incidents. we may want to include extra incident fields for contest_id, cycles and metrics

**Comment by na-ji** [comment]
Regarding the contest entries, I would say that only the top 3 player's scores would be useful. Since only the top 3 players are getting the most generous rewards, people might want to get a spot on those. So I do agree with @jfberry that a flatter structure would be easier here.

**Comment by jfberry** [comment]
It may then make sense to have this in a json field on pokestop if the content values are unlikely to be queried directly (ie only need to be parsed to be included in a webhook and as a whole entity for a map).

**Comment by jfberry** [comment]
That was why I was thinking having a json field with contest results on pokestop (ie only one contest available at a time, but format is extensible to later contests).  Right now we could populate with top-3 in there.  This gives us the chance to store and retrieve data without becoming too opinionated just yet.  If contests evolve or the data is used for things we don't yet imagine we can break out the data appropriately.

**Comment by TurtIeSocks** [comment]
So the following 3 columns?
- Contest type
- Total # of entries
- Information on top 3 trainers ( if we're storing in a json it could just be an array and have it be configurable? )

**Comment by na-ji** [comment]
>     * Contest type
It's actually two values in the protos: 
```
    pokemon_metric ContestPokemonMetric.POKEMON_SIZE:1
    ranking_standard ContestRankingStandard.MAX:2
```

> Do we need total number of entries? that will change more frequently than just the top 3? Not opposed, but if it has no value then not worth a write.

There are a limited number of participants in contests. So it can be useful to know whether the contest is full or not.

**Comment by ccev** [comment]
realistically there's more interesting values, like the focus pokemon id (and whatever other focus values there are?) and I believe the max amount of players can also vary per contest. there's probably many more values that can become relevant based on how contests will evolve.

I worry that the current idea for structure will become cumbersome and backstab us in the future. I feel like we know too little about contests and where they're going with them.

Personally, I'd refrain from storing the top 3 for now and only go for the json. For every top 3 entry, you'd need the whole pokemon display, the score, the name and whatever else there is. Realistically you'll never need these values unless you want to display them on a map, so storing them as their individual fields will just be unnecessary bloat. Just scoring the top 3 scores (or only the top 1 score) is probably enough, and the entire entry list as JSON.

**Comment by na-ji** [comment]
> realistically there's more interesting values, like the focus pokemon id (and whatever other focus values there are?)

True, I forgot about this one.


> I believe the max amount of players can also vary per contest

For now this is a global setting in the GM, CF pokeminers: 
![image](https://github.com/UnownHash/Golbat/assets/757637/0809d766-dd83-4f40-a4d3-cee02e32bf55)



> Personally, I'd refrain from storing the top 3 for now and only go for the json. For every top 3 entry, you'd need the whole pokemon display, the score, the name and whatever else there is. Realistically you'll never need these values unless you want to display them on a map, so storing them as their individual fields will just be unnecessary bloat (I count 8*3 columns). Just scoring the top 3 scores (or only the top 1 score) is probably enough, and the entire entry list as JSON.

This is the reason I proposed to only save the top 3 players scores. We don't care about other's values: the Pokémon is the same for everybody if there is a focus, and when there isn't it won't change anything. We don't want personal information about players, either. The top 3 score is interesting because the best rewards are given to the top 3 players. 



> I worry that the current idea for structure will become cumbersome and backstab us in the future. I feel like we know too little about contests and where they're going with them.

Agreed. There is also the `schedule` property that seems to hold useful information. Maybe contest can be recurrent or this kind of thing.

**Comment by jfberry** [comment]
This is exactly why I suggested using a json field (and a json field only); keep it nice and compact with what we really care about now - but it's easily expandable later as we learn more.
I don't think we should just store everything as this is just bloat. But we should be prepared to build out as we understand the use case more

**Comment by ccev** [comment]
@jfberry how do you feel about a separate contest table that's 1:1 related with incident? Just to keep all these extra fields out of the incident table

**Comment by jfberry** [comment]
I'm not proposing any extra fields on incident. The incident id is not even in these protos as far as I can see.
I'm proposing a single json field on pokestop.

**Comment by na-ji** [comment]
> I'm not proposing any extra fields on incident. The incident id is not even in these protos as far as I can see. I'm proposing a single json field on pokestop.

There is an incident_id:
```
pokestop_displays {
    incident_id "-1601456624153831644"
    incident_start_ms 1688626800000
    incident_expiration_ms 1688666400001
    incident_display_type IncidentDisplayType.INCIDENT_DISPLAY_TYPE_INCIDENT_CONTEST:9
}
```

Also, when using `METHOD_GET_CONTEST_DATA`, the game returns a repeated Contest proto, so we might have multiple contests per Pokestop in the future, no?

**Comment by ccev** [comment]
i think there were multiple Niantic devs involved with contests that didn't properly communicate with eachother. 

There's a Incident ID, that seems to be only used in the GMO. Then there's a Contest ID, that's only used to get entries. When interacting with contests in any other way (i.e. placing a pokemon), you use the Fort ID + Lat + Lon. So overall there's three ways they use to ID contests.

The part I shared in this issue 100% supports multiple contests on a single Pokestop. But all other requests only support a single contest on a Pokestop.

I think that this should be stored in a 1-n relation to Pokestops, just to be safe

**Comment by TurtIeSocks** [comment]
Perhaps instead we should focus on discussing *how* this data will be used by the frontends and that will perhaps lead to us to an agreement on which columns/data we want to save.

**Comment by Fabio1988** [comment]
Talked with my community, most important information would be the following:

> Showcase largest Relaxo
> runs until 15.08.2023 until 20:00 clock
> Rank 1 (912)
> Rank 2 (889)
> Rank 3 (864)
> Total participants 8

Not sure where we should store that, we can store it on pokestop table in a json. But there could be multiple contests in the proto, is that related to multiple stops or could they be on same stop?

**Comment by Mygod** [comment]
For map, we need to have filtering for stops where your best mon having x points can make it into top 3, so either if there are <3 entries or points of rank 3 is <= x. Pokemon species, largest/smallest, end time is also nice to have although all stops seem to share those for the past contests. Not sure if # of participants matter at all though.

**Comment by Fabio1988** [comment]
https://www.reddit.com/r/TheSilphRoad/comments/152yg6k/showcases_biggest_points_formula_silph_research/jsgrxmn

**Comment by jfberry** [comment]
The score for Pokémon in the showcase is in the proto anyway, and maps can’t calculate scores of wild Pokémon size exact weight and height no longer known.

**Comment by Mygod** [comment]
> > ```
> > * Contest type
> > ```
> 
> It's actually two values in the protos:
> 
> ```
>     contest_metric {
>         pokemon_metric ContestPokemonMetric.POKEMON_SIZE:1
>         ranking_standard ContestRankingStandard.MAX:2
>     }
> ```

Actually `pokemon_metric` can only be `ContestPokemonMetric.POKEMON_SIZE` or `UNSET` as of now.

I propose the following 7 fields, with the first field being a JSON.

```
contest_info {
    "focus": {
        "pokemon": 7,
        "form": null
    },
    "metric": "max-size",
    "cycle": {
        "start": 1688630400000,
        "occurrence": "custom",
        "warmup": 3600000,
        "cooldown": 1
    },
    "updated": scanned_time_ms
}
contest_end_ms 1688666400000
contest_entry_count 13
contest_score_1 900.1
contest_score_2 899.9
contest_score_3 696.9
contest_score_updated entries_scanned_time_ms
```

Arguably we can replace the `contest_score_updated` column with a JSON field storing a more detailed table of contest entries (which goes beyond top 3). Example:

```
contest_entries {
    "entries": [
        {"pokemon", "form", "costume", "score", "trainer_name", "team", ...},
        ...
    ],
    "updated": entries_scanned_time_ms
}
```

Thoughts?

**Comment by jfberry** [comment]
1. No other fields are in ms/all are converted to seconds
2. I think we decided that we would just have a single json field for everything for now since we don't actually need to query on any of this data (though I'm ambivalent on this point)
3. Is cycle valuable? 
4. If the number of entries doesn't matter (I don't think it does) then not storing this means that we don't get pointless updates

**Comment by Mygod** [comment]
1. We should not unnecessarily postprocess data. This literally wastes a few cpu cycles for division and has zero benefits whatsoever. No need to try to maintain RDM compatibility. #everymillisecondmatters
2. As I've already said, map needs filtering. https://github.com/UnownHash/Golbat/issues/125#issuecomment-1638574265
3. Contests can be daily.
4. It does if you have happened to read any one of these comments: https://github.com/UnownHash/Golbat/issues/125#issuecomment-1624950107 https://github.com/UnownHash/Golbat/issues/125#issuecomment-1625245807 https://github.com/UnownHash/Golbat/issues/125#issuecomment-1625359437 https://github.com/UnownHash/Golbat/issues/125#issuecomment-1638574265

**Comment by jfberry** [comment]
1. Avoiding a divide by 1000 is not more important than being internally consistent. It also requires 64-bit storage and all other times use 32-bit storage taking up half the space on disk for each row. I'll leave for others to weigh in, but I don't think this works.
2. This has index implications also in this case. Arguably only the third place would be needed for an index though. Interested in a view from @TurtIeSocks on what level of filtering he would add to Reactmap
3. So what is the implication? What would you do with cycle? It's just taking up storage space if it is not usable (and can be added later)
4. Fine

**Comment by Mygod** [comment]
Re 1: We can add _ms to the suffix to discriminate from legacy seconds and help readability.

Re 2: There's probably no need to do indexing on them, then I guess it is debatable whether we need separate fields at all or should we just have SQL extract from JSON in WHERE queries. I guess if we do not do index, then might as well just do two json fields `contest_info` and `contest_entries`.

Re 3: Judging from the code, it seems that a daily contest has three periods: warmup, active, cooldown. Only in the second phase you can put in your mon. It would be helpful to know when a contest is active.

**Comment by jfberry** [comment]
1. Either we use seconds consistently or we agree the codebase should be translated to milliseconds everywhere over time (which would be more consistent with the protos). Having some things in one format and some in another doesn't make sense. (The original decision to truncate to seconds was made by RDM developers and we have maintained compatibility up until now. Perhaps this was originally a storage size concern, but I doubt it - and if we were to transition we would likely need to maintain both in parallel for some time for down-stream tooling)
2. extracting from JSON in where clause is excruciatingly slow, especially where tables are large - as these are. You would certainly want an index if you wanted to filter on this via sql. This is why I asked Turtle what he intended to do for initial implementation - since reactmap is currently already very slow in this area.
3. So how would these fields be used? Since these contests don't exist yet perhaps these should be added when they are understood.  Perhaps malte knows

**Comment by Mygod** [comment]
Re 1: I am precisely proposing to migrating to ms everywhere. Using seconds as int32 timestamps would eventually run into overflow issue in a few years. If we use int64, might as well keep the original ms.

Re 3: We can leave out the cycle for now and add it back when the game actually runs a non-custom-cycle contest.

**Comment by Fabio1988** [comment]
> Re 1: I am precisely proposing to migrating to ms everywhere. Using seconds as int32 timestamps would eventually run into overflow issue in a few years. If we use int64, might as well keep the original ms.
> 
> Re 3: We can leave out the cycle for now and add it back when the game actually runs a non-custom-cycle contest.

You are really thinking of a use case in some years? Converting to seconds would lead to inconsistency to RDM and all other tools.... I would prefer the way of seconds.... But useless for that discussion, we keep seconds

**Comment by Mygod** [comment]
Fun fact: Year 2038 problem is only less than 1.5 decades away! Very scary!!!

Personally I'd like to keep existing fields as seconds and only use ms for new features/fields to keep compatibility with downstream. (Downstream is going to write new code to support this anyways so they might as well adapt.) Some of Chuck also uses ms for new feature timestamping.

**Comment by TurtIeSocks** [comment]
> 1. Avoiding a divide by 1000 is not more important than being internally consistent. It also requires 64-bit storage and all other times use 32-bit storage taking up half the space on disk for each row. I'll leave for others to weigh in, but I don't think this works.
> 2. This has index implications also in this case. Arguably only the third place would be needed for an index though. Interested in a view from @TurtIeSocks on what level of filtering he would add to Reactmap
> 3. So what is the implication? What would you do with cycle? It's just taking up storage space if it is not usable (and can be added later)
> 4. Fine
> 

@jfberry looking at the data, the only fields that come to mind for filtering this early on are the focus Pokemon/form, metric, and end time. 

Even then, it might be a bit early since metric only has one type and Niantic hasn't put out contests with multiple Pokemon IDs yet.

**Comment by Mygod** [comment]
New event is holding contests that showcases a different Pokemon for each stop, so this has to be in a column as well. 😢
